File: 6. Data Structures Arrays/9.MergeSortedArrays.py
# Python - 3.10
# Topic - Interview Question - Data Structures and Algorithms
# program - Given 2 arrays that are sorted, merge this two arrays into one big one

# Input
# [0, 3, 4, 31], [4, 6, 30]
# Output
# [0, 3, 4, 4, 6, 30, 31]


# Starting Interview

# You - So does my result have to be sorted
# interviewer - Ya, it has to be

# 1ST APPROACH
# My first approach would be to create a new list in my function as well as an i,j = 0; variables that i will use in a while loop to check and compare each item in both arrays, if both items are equal to i want to add both and increment i, j, if one is less than the other i want to add the lesser one and increment it's index for the next comparison. At the point where we have exhausted any of our array we want to simply add the remainder of the other array and then break out of the loop

# interviewer: i love that. let's see that in code


def mergeSortedArrays(arr1, arr2):
   # Input Validation
   if (len(arr1) == 0 or len(arr2) == 0): return arr1 + arr2
   bigList = []
   i = 0
   j = 0

   # [0, 3, 4, 31]   | len = 4
   # [4, 6, 30]  | len = 3
   while True:
      # check if we've exhausted any array, add the other array remaining and break out of loop
      if (i == len(arr1) and j == len(arr2)): 
         break
      elif (i == len(arr1)):
         bigList += arr2[j:]
         break
      elif (j == len(arr2)):
         bigList += arr1[i:]
         break

      if (arr1[i] == arr2[j]):
         bigList.append(arr1[i])
         bigList.append(arr2[j])
         i += 1
         j += 1
      elif (arr1[i] < arr2[j]):
         bigList.append(arr1[i])
         i += 1
      elif (arr1[i] > arr2[j]):
         bigList.append(arr2[j])
         j += 1

   # print(f"i: {i} | j: {j}")   
   print(bigList)
   return bigList


arr1, arr2 = [0, 3, 4, 31], [4, 6, 30]

mergeSortedArrays(arr1, arr2)


# Interviewer - That's an absoultely great solution. How would you make this code more readable and cleaner

File: 6. Data Structures Arrays/9.MergeSortedArrays.js
// Javascript
// Topic: Interview Question - Data Structures & Algorithm
// Program: Given 2 arrays that are sorted, merge this two arrays into one big one

// Input
// [0, 3, 4, 31], [4, 6, 30]
// Output
// [0, 3, 4, 4, 6, 30, 31]

function mergeSortedArrays(arr1, arr2){
   // Check Input
   if (arr1.length === 0 || arr2.length === 0){
      arr1.push(...arr2)
      return arr1;
   }
   
   let mergedArray = [];
   let i = 0; 
   let j = 0;

   // Length of arrays
   let [arr1Length, arr2Length] = [arr1.length, arr2.length]
   
   while (i < arr1Length || j < arr2Length) {
      // comparison values initialization
      let arr1item = arr1[i];
      let arr2item = arr2[j];

      // Check: to see if we have added all the elements of either arr1 or arr2
      [mergedArray, done] = checkArryays(mergedArray, arr1, arr2, arr1Length, arr2Length, i, j);
      if (done) break;

      // Merge Arrays: add items from both arrays in a sorted order
      [mergedArray, i, j] = mergeArrays(mergedArray, arr1item, arr2item, i, j);
   }
   console.log(mergedArray)
}

const arr1 = []; 
const arr2 = [5, 6, 7];

mergeSortedArrays(arr1, arr2)

function checkArryays(mergedArray, arr1, arr2, arr1Length, arr2Length, i, j) {
   // Program is done variable
   let done = false;

   if (i === arr1Length){
      mergedArray.push(...arr2.slice(j,));
      done = true;
   } else if (j === arr2Length){
      mergedArray.push(...arr1.slice(i,));
      done = true;
   }
   return [mergedArray, done]
}

function mergeArrays (mergedArray, arr1item, arr2item, i, j) {
   if (arr1item === arr2item){
      mergedArray.push(arr1item);
      mergedArray.push(arr2item);
      i++;
      j++;
   } else if (arr1item < arr2item) {
      mergedArray.push(arr1item);
      i++;         
   } else if (arr1item > arr2item) {
      mergedArray.push(arr2item);
      j++;
   }

   return [mergedArray, i, j]
}